doc: improve readme

dvrogozh opened the pull request:
Fixes: #1731

disable_all

CC: @gajanan-choudhary, @EikanWang

gajanan-choudhary commented on README.md:
```diff
@@ -1,14 +1,20 @@
 # Torch XPU Operators*
 
-Torch XPU Operators* implements PyTorch ATen operators for Intel GPU devices, aiming to agilely support PyTorch ATen operations and buffer these operations for Intel GPU upstreaming . For more details, refer to [SYCL kernels for ATen Operators RFC](https://github.com/pytorch/pytorch/issues/114835) for more details.
+Torch XPU Operators* project is integral part of [PyTorch](https://github.com/pytorch/pytorch) to support XPU acceleration backend. PyTorch build system automatically clones this repository at the pin pointed commit, branch or tag specified in the following file:
```
```suggestion
Torch XPU Operators* project is integral an part of [PyTorch](https://github.com/pytorch/pytorch) to support XPU acceleration backend. The PyTorch build system automatically clones this repository at the commit, branch, or tag pinned and specified in the following file of the [PyTorch repository](https://github.com/pytorch/pytorch):
```

gajanan-choudhary commented on README.md:
```diff
@@ -1,14 +1,20 @@
+* PyTorch ATen operators
+* Torchvision operators
+
+Most operators are implemented as SYCL kernels sources of which are available within this repository. Some operators (linear algebra)
+are implemented thru the calls to [oneMKL](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html#gs.mdl4z3) library.
```
```suggestion
Most operators are implemented as SYCL kernels, the sources of which are available in this repository. Some operators (linear algebra)
are implemented through calls to the [Intel® oneAPI Math Kernel Library (oneMKL)](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html).
```

gajanan-choudhary commented on README.md:
```diff
@@ -1,14 +1,20 @@
+Most operators are implemented as SYCL kernels sources of which are available within this repository. Some operators (linear algebra)
+are implemented thru the calls to [oneMKL](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html#gs.mdl4z3) library.
+
+Note that few operators (convolution and gemm) for XPU backend are implemented directly in PyTorch sources thru the calls to [oneDNN](https://github.com/uxlfoundation/oneDNN) library. These sources can be found at https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native/mkldnn/xpu folder.
```
```suggestion
Note that a few operators (convolution and matrix-matrix multiplication (`gemm`)) for the PyTorch XPU backend are implemented directly in PyTorch sources through calls to the [oneAPI Deep Neural Network Library (oneDNN)](https://github.com/uxlfoundation/oneDNN). These sources can be found in the PyTorch repository at https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native/mkldnn/xpu.
```

gajanan-choudhary commented on README.md:
```diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
-## Build
+## Build and install
+
+This project can't be built or installed stand alone. Project is used when PyTorch is built with XPU backend support.
```
```suggestion
This project cannot be built or installed as a stand-alone. This project gets built when PyTorch is built with XPU backend support.
```

gajanan-choudhary commented on README.md:
````diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/xpu
+```
+
+**To build PyTorch with XPU backend from sources**, refer to PyTorch [documentation](https://github.com/pytorch/pytorch/blob/main/README.md#from-source) and pay attention on [Intel GPU Support](https://github.com/pytorch/pytorch/blob/main/README.md#intel-gpu-support) section. Shortly, PyTorch build can be triggered as follows:
````
```suggestion
**To build PyTorch with XPU backend from sources**, refer to the [Intel GPU Support](https://github.com/pytorch/pytorch/blob/main/README.md#intel-gpu-support) section of the [PyTorch documentation](https://github.com/pytorch/pytorch/blob/main/README.md#from-source). In summary, the PyTorch build for XPU backend can be triggered as follows:
```

gajanan-choudhary commented on README.md:
````diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+--   USE_XCCL              : ON
 ```
 
+If building from sources, note the following environment variables which control PyTorch XPU backend build:
````
```suggestion
If building from sources, note the following environment variables which control the PyTorch XPU backend build:
```

gajanan-choudhary commented on README.md:
````diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+python setup.py install >log.txt 2>&1
+```
+
+Monitor for the following log printout indicating that XPU support is being built:
````
```suggestion
Look for the following lines in the log file indicating that XPU backend is being built:
```

gajanan-choudhary commented on README.md:
```diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+| `USE_XCCL` | `ON` (>= PT2.8) | Enables XCCL distributed backend support |
+| `TORCH_XPU_ARCH_LIST` | depends on the PT and OS versions | Build SYCL kernels for specified platform(s) |
+
+The `TORCH_XPU_ARCH_LIST` allows to specify comma separated list of platforms for which SYCL kernels will be built. This helps to reduce build time. Option does not affect oneDNN or oneMKL based operators. Note that if PyTorch will be executed on the platform for which SYCL kernels were not built, they will be JIT compiled.
```
```suggestion
The `TORCH_XPU_ARCH_LIST` allows to specify a comma separated list of platforms for which SYCL kernels will be built. This helps to reduce the build time. This option does not affect oneDNN- or oneMKL-based operators. Note that if PyTorch is executed on a platform that the SYCL kernels were not built for, then they are compiled just-in-time (JIT).
```

gajanan-choudhary commented on README.md:
```diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+
+The `TORCH_XPU_ARCH_LIST` allows to specify comma separated list of platforms for which SYCL kernels will be built. This helps to reduce build time. Option does not affect oneDNN or oneMKL based operators. Note that if PyTorch will be executed on the platform for which SYCL kernels were not built, they will be JIT compiled.
+
+## Verify
```
```suggestion
## Verification
```

gajanan-choudhary reviewed (changes requested):
Thanks for the quick response to #1731. Some fixes are needed, ready to approve once they are addressed.

dvrogozh commented:
@gajanan-choudhary : I addressed most of the comments (changes pushed), but still work on the few. Will update again today.

dvrogozh commented on README.md:
```diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+
+**Does IPEX uses PyTorch XPU backend implementation?**
+
+Yes. IPEX relies on PyTorch XPU backend implementation and augments it with additional features and operators. On top of that IPEX implements select features and operators outside of standard PyTorch API which are required in popular AI frameworks such as vLLM, Huggingface TGI, SGLang, and others. The ultimate long term goal is to upstream or substituite with better upstream implementations as ma ...
```
I will drop last 2 sentences as they indeed might be confusing and will require to add too much text to thoroughly explain.

> I also did not understand the last sentence of this answer. Could you clarify?

There are IPEX specific APIs which are getting replaced by PyTorch APIs. Distributed backend is one of such examples. It changes a name - in IPEX its analogues was called "ccl", in pytorch it's "xccl". And there are other API level differences in its usage.

gajanan-choudhary replied:
I see, I understand better now. I'm okay with dropping the last two sentences. Feel free to resolve this thread (I don't have the option to resolve).

dvrogozh commented:
> I addressed most of the comments (changes pushed), but still work on the few. Will update again today.

All done from my side. Please, review again.

gajanan-choudhary commented on README.md:
````diff
@@ -17,15 +23,80 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+
+```
+$ cat log.txt | grep -E "(USE_XPU|USE_XCCL)\s*:"
+--   USE_XPU               : 1
+--   USE_XCCL              : ON
````
Thanks for the clarification. We can keep it as is then.

dvrogozh commented:
@gajanan-choudhary, if you are ok with the last version, can you, please, approve PR?

EikanWang commented on README.md:
````diff
@@ -17,14 +27,78 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+True
+```
+
+# FAQ
````
From my understanding, the FAQ should reside within the IPEX repository, and there's no need to explicitly link it to IPEX elsewhere. 

By the way, I'm not entirely clear on IPEX's current positioning—could you please confirm this with the IPEX owner?

dvrogozh commented:
@EikanWang : are you ok to merge this version?

riverliuintel commented on README.md:
```diff
@@ -17,14 +27,75 @@ For the hardware and software prerequiste, please refer to [PyTorch Prerequisite
+
+# FAQ
+
+Some confusion might arise about relationship of this repostory and PyTorch XPU backend in general with [Intel® Extension for PyTorch (IPEX)](https://github.com/intel/intel-extension-for-pytorch). See answers on that below.
```
IPEX is not an valid name, please use full name, not IPEX. And Intel Extension for PyTorch is deprecating, even not highlight the question is fine here.

chuanqi129 commented:
Hi @dvrogozh please help to address @riverliuintel 's comments, after that we can land this PR

dvrogozh commented:
> Hi @dvrogozh please help to address @riverliuintel 's comments, after that we can land this PR

Addressed, dropped IPEX abbreviation.

riverliuintel commented:
Looks good
